### PR TITLE
Allow to disable the password-based login (sign-in) form

### LINKED ISSUE #7633: Is it possible to choose default auth source?
- Gitea version (or commit ref): 1.10.0+dev-60-gc5f9c0418
- Git version:
- Operating system: linux
- Database (use `[x]`):
  - [ x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x ] Not relevant

### LINKED ISSUE #13606: Allow to disable local login
- Gitea version (or commit ref): 1.12.6
- Git version: N/A
- Operating system: N/A
  
  
  
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist: N/A

## Description
When using gitea in an environent (e.g. corporate deployment) when SSO login is required, it would be nice to have a way to disable local login, to not have a confusing login-page, which can't be used.

A nice extra would be that, when only one authentication-source is configured, clicking on the "Sign in" button on the header automatically redirects to the IdP/SSO.

### PR DESCRIPTION
**Pull request by wxiaoguang**
Usually enterprise/organization users would like to only allow OAuth2 login.

This PR adds a new config option to disable the password-based login form. It is a simple and clear approach and won't block the future login-system refactoring works.

https://gitea.com/gitea/docs/pulls/110

Fix a TODO in #24821

Replace  #21851

Close #7633 , close #13606

Screenshots:

<details>

image

image

</details>

### CODE REVIEW

File: routers/web/auth/auth.go [hunk 1]
```diff
@@ -160,54 +160,42 @@ func CheckAutoLogin(ctx *context.Context) bool {
...
 }
 
 // SignInPost response for sign in request
 func SignInPost(ctx *context.Context) {
-	ctx.Data["Title"] = ctx.Tr("sign_in")
-
-	oauth2Providers, err := oauth2.GetOAuth2Providers(ctx, optional.Some(true))
-	if err != nil {
-		ctx.ServerError("UserSignIn", err)
+	if !setting.Service.EnablePasswordLoginForm {
```
Comment (delvh) [inline routers/web/auth/auth.go]: By the way, doesn't this imply that `DISABLE_REGISTRATION` is `true` as well?
Or what's the use of offering a sign-up page where you can set a password that you cannot use afterward?
Perhaps we should add a warning in the logs when `!setting.Service.EnablePasswordLoginForm && !DisableRegistration`.

**Review by delvh — APPROVED** [review]

File: routers/web/auth/auth.go (hunk 1 above)
Comment (wxiaoguang) [reply routers/web/auth/auth.go]: No, that's unrelated.

An account could be created manually with DISABLE_REGISTRATION=true, and then make it login by password

Comment (delvh) [reply routers/web/auth/auth.go]: I think we are talking about different points:

>  and then make it login by password

How if the password login is disabled?

I am not talking about the case `registration is disabled and thus login should be disabled`, I'm talking about the other way around:
`password login is disabled thus (password) registration should be disabled`

Comment (wxiaoguang) [reply routers/web/auth/auth.go]: Hmm, you reminded me that we should use the same wording ....

renamed all names from "login" to "signin" in bb7f05424aa55e73ded56ec99406429eeabe83d7

Comment (wxiaoguang) [reply routers/web/auth/auth.go]: > How if the password login is disabled?

But ENABLE_BASIC_AUTHENTICATION could still be true.

Comment (delvh) [reply routers/web/auth/auth.go]: Wait, now I'm confused:
What's the difference between `ENABLE_PASSWORD_SIGNIN_FORM` and `ENABLE_BASIC_AUTHENTICATION`?
Where else can you still use your password?
For the API (where IIRC it is deprecated for removal) or `git push`?
Or somewhere else?

Comment (wxiaoguang) [reply routers/web/auth/auth.go]: > For the API (where IIRC it is deprecated for removal) or `git push`?

yes, including git-push and API.

----

> where IIRC it is deprecated for removal

But it it still there.

image

**Review by lunny — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Need to fix a regression:  Fix oauth2 login methods #32698

**Pull request merged**